### PR TITLE
 fix #1110: add /user/breach-stats endpoint 

### LINKED ISSUE #1110: Add /user/getBreachStats endpoint
It accepts an OAuth access token in the `Authorization: Bearer` header. (The access token should have a `monitor` scope.)

If the access token is for a user who has an active Monitor subscription, return a JSON summary of their breach statistics:
* number of Email addresses being monitored
* number of data breaches that have exposed their data
* how many times their password(s) have been exposed

See [invision spec](https://mozilla.invisionapp.com/d/main#/console/17296602/360923276/preview).

### PR DESCRIPTION
**Pull request by groovecoder**
For #1110, I want to get this up for review ASAP.

Still TODO:
* [x] Change endpoint to accept OAuth token in `Authorization` HTTP header instead of url parameter

### CODE REVIEW

File: controllers/user.js [hunk 1]
```diff
@@ -337,6 +338,20 @@ async function getPreferences(req, res) {
 }
 
 
+async function getBreachStats(req, res) {
+  const fxaResponse = await FXA.verifyOAuthToken(req.query.fxa_oauth_token);
```
Comment (pdehaan) [inline controllers/user.js]: I'm guessing if an `fxa_auth_token` wasn't passed via the querystring, that'd just be `null` or `undefined`. Not sure if we'd need to do a check for a valid-looking oauth token before proceeding or if that is just overkill.

File: controllers/user.js [hunk 2]
```diff
@@ -337,6 +338,20 @@ async function getPreferences(req, res) {
...
+  const user = await DB.getSubscriberByFxaUid(fxaResponse.body.user);
```
Comment (pdehaan) [inline controllers/user.js]: Should we do any checks here to verify that `user` found a match?

File: scan-results.js [hunk 3]
```diff
@@ -95,4 +95,52 @@ const scanResult = async(req, selfScan=false) => {
...
+    },
+    passwords: {
+      subhead: "",
+      affectedAccounts: {},
+      count: 0,
+    },
+  };
+  let foundBreaches = [];
+
+  breachStats.monitoredEmails.count = verifiedEmails.length;
```
Comment (pdehaan) [inline scan-results.js]: Why not set this on L102 above?

File: scan-results.js [hunk 4]
```diff
@@ -95,4 +95,52 @@ const scanResult = async(req, selfScan=false) => {
...
+  breachStats.monitoredEmails["allVerifiedEmails"] = verifiedEmails;
```
Comment (pdehaan) [inline scan-results.js]: Same here, why not set this in the initial object above?

File: controllers/user.js (hunk 1 above)
Comment (groovecoder) [reply controllers/user.js]: Added checks and appropriate responses.

**Comment by groovecoder** [comment]
Addressed all comments and updated the code to get the token from `Authorization:` header instead of URL parameter.

**Comment by groovecoder** [comment]
Updated to fix tests too. @lesleyjanenorton r?

File: controllers/user.js [hunk 5]
```diff
@@ -337,6 +338,35 @@ async function getPreferences(req, res) {
...
+    });
+  }
+  const fxaResponse = await FXA.verifyOAuthToken(req.token);
+  if (!fxaResponse) {
+    return res.status(404).json({
+      errorMessage: "Cannot find FXA for that OAuth token.",
+    });
+  }
+  const user = await DB.getSubscriberByFxaUid(fxaResponse.body.user);
+  if (!fxaResponse) {
```
Comment (pdehaan) [inline controllers/user.js]: I think this is unreachable sine we check-n-bail above on L348.
Should this be `if (!user)`?

Comment (groovecoder) [reply controllers/user.js]: Oh, yes. Duh! Good catch.

**Pull request merged**